### ISSUE TITLE
Add support to TLAPS

### ISSUE THREAD

**Issue by andylokandy** [alygin/vscode-tlaplus#153]
Is it possible to have `vscode-tlaplus` run tlapm to check the proofs and even to decompose the proofs?

**Comment by lemmy** [comment]
Here is a high-level description of what is needed:

Must
---
- Extension settings to set tlapm path
- Invoke tlapm from vscode extension
- Parse tlapm output and annotate editor accordingly

Should
----
- Pipe tlapm to a console
- Dialog to configure tlapm flags per invocation
- Editor coloring similar to Toolbox
- Refactor and extract decompose proof command out of Toolbox into standalone library
- Integrate decompose proof command in extension

Nice to have
----
- Settings to change color mapping

**Comment by dricketts** [comment]
Proof folding and the associated hiding commands from the Toolbox would also be nice to have.

**Comment by lemmy** [comment]
@dricketts It occurred to me that I would find it useful if all facts were visually annotated in the module that appear in the ```BY DEF``` statements of the current proof step.  I hope that it would help make it easier to spot missing facts.  What's your take?

**Comment by dricketts** [comment]
@lemmy I'm not sure I understand what you mean. Suppose a step has `BY <2>3, Lemma1 DEF Foo`. Would this feature visually annotate step `<2>3`, `Lemma1`, and `Foo`?

**Comment by lemmy** [comment]
It would annotate the *definition* of `Lemma1` and `Foo` where they appear in the spec.

**Comment by dricketts** [comment]
I don't think the annotations would help me debug a proof step. What I usually miss is a way to drop into a procedural proof mode (like tactic-based proofs in Coq) in order to understand what's missing. This wouldn't be a replacement for structured TLA+ proofs - procedural proofs are unreadable and difficult to maintain. But I think it would be a faster way to figure out what definitions and facts are missing, or whether the proof just needs to be broken down further.

Of course, such a mode is much more difficult to implement than annotating definitions. So if others think that feature would be valuable, I'm all for it.

**Comment by will62794** [comment]
I was recently working on some TLAPS proofs and it would have been nice to have this feature. Even if it was only able to check proofs and nothing else. I started poking around a bit to see what would be needed to implement a minimum viable solution in the VS Code extension. I suppose @dricketts would have some of the best insights here since it seems he implemented a good portion of the initial TLAPS functionality in the Toolbox, if my understanding is correct. One thing I am curious about, which seems a potentially larger hurdle to straightforward implementation, is whether coloring of proof steps in the Toolbox requires use of the SANY parse tree. From looking at a [few files](https://github.com/tlaplus/tlaplus/blob/875bf8ffc1a850cdf76b142d7c8c40d70f59b0c4/toolbox/org.lamport.tla.toolbox.tool.prover/src/org/lamport/tla/toolbox/tool/prover/ui/output/data/ColorPredicate.java) it seems that it does, but I am not certain based only on a cursory reading of the code. If this is the case, I was wondering if the new [tree-sitter parser](https://github.com/tlaplus-community/tree-sitter-tlaplus) from @ahelwer would be useful here, if we need richer information about the parse tree to implement proof step coloring properly inside the VSCode extension. As I understand it, SANY does have some kind of [XML export feature](https://github.com/tlaplus/tlaplus/blob/875bf8ffc1a850cdf76b142d7c8c40d70f59b0c4/tlatools/org.lamport.tlatools/src/tla2sany/xml/XMLExporter.java), but I don't know if it is maintained, or if it works for TLAPS proofs.

**Comment by ahelwer** [comment]
@will62794 yeah the tree-sitter grammar properly parses proofs with levels and so forth. Proof folding/hiding is implemented in Neovim already through the nvim-treesitter plugin. What is this coloring you refer to? Like regular syntax highlighting? Or coloring to mark a proof step as proven or not?

**Comment by will62794** [comment]
@ahelwer Yeah, the tree-sitter grammar looks awesome. I tried out the [online demo](https://tlaplus-community.github.io/tree-sitter-tlaplus/) and it seems to parse TLAPS proofs just fine (e.g. a basic [sample proof](https://raw.githubusercontent.com/will62794/logless-reconfig/master/proofs/BasicQuorumsLib.tla)). It seems you already have a [demo](https://github.com/tlaplus-community/tlaplus-tool-dev-examples/tree/2ea5bdb480229f6f419154cb3ff319bd8f0b99c9/tree-sitter/typescript) that works in TypeScript, so presumably it could be used directly from within the VS Code extension, if needed.

> What is this coloring you refer to? Like regular syntax highlighting? Or coloring to mark a proof step as proven or not?

The latter i.e. the coloring the Toolbox shows to indicate whether a proof step is "proved", "failed", "in progress", etc. To my knowledge, the [TLA+ proof manager](https://github.com/tlaplus/tlapm) handles proving the necessary proof obligations and reporting the status back to the Toolbox, but the information it reports doesn't seem sufficient to generate the coloring information that is currently shown in the Toolbox. There seems to be a more involved computation for determining the coloring of proof steps based on the results of the proof obligations returned by TLAPM e.g. see [here](https://github.com/tlaplus/tlaplus/blob/875bf8ffc1a850cdf76b142d7c8c40d70f59b0c4/general/docs/ProofStatus.tla). Part of this computation seems to rely on knowledge of the syntax tree. I will let @dricketts or @lemmy correct me if any of this is inaccurate.

Also, I suppose the extension does already [have some parsing capabilities](https://github.com/alygin/vscode-tlaplus/blob/2092aeb3e1a006613c92caef668ae87e42a5d293/languages/tlaplus-grammar.json), which it seems to rely on for syntax highlighting, but it may be too basic to provide detailed information about TLAPS proof trees.

**Comment by lemmy** [comment]
IIRC the proof coloring is exclusively implemented in the Toolbox. The same goes for the proof decomposition functionality.  Perhaps more of this functionality should be moved out of the Toolbox and pushed downwards into, e.g., TLAPS.

/cc @johnyf

**Comment by kape1395** [comment]
Do you know if there is any progress on this?
I want to work on the TLAPS integration this summer and wonder if that overlaps with others.

**Comment by lemmy** [comment]
@kape1395 Thanks for your contribution! I'm not aware of any other efforts to add TLAPS support to VSCode. Would you like to join the upcoming TLA+ project meeting on 6/13 at 8 am PDT to collaborate with others if any?

/cc @muenchnerkindl

**Comment by lemmy** [comment]
@kape1395 Is it time to close this issue? :-)

**Comment by kape1395** [comment]
@lemmy, some support is in place, but the feature parity is not here yet (proof decomposition).
Can we close this issue and open new issues for the proof decomposition and other aspects?
The former should be in the TLAPM repo because the implementation will be there.
